From: Uwe Kleine-König <uwe@kleine-koenig.org>
Subject: [PATCH] rtc: s35390a: make sure all members in the output are set
Forwarded: not-needed

The rtc core calls the .read_alarm with all fields initialized to 0. As the
s35390a driver doesn't touch some fields the returned date is interpreted as
a date in January 1900. So make sure all fields are set to -1; some of them
are then overwritten with the right data depending on the hardware state.

In mainline this is done by commit d68778b80dd7 ("rtc: initialize output
parameter for read alarm to "uninitialized"") in the core. This is
considered to dangerous for stable as it might have side effects for other
rtc drivers that might for example rely on alarm->time.tm_sec being
initialized to 0.

Signed-off-by: Uwe Kleine-König <uwe@kleine-koenig.org>
---
--- a/drivers/rtc/rtc-s35390a.c
+++ b/drivers/rtc/rtc-s35390a.c
@@ -267,6 +267,20 @@
 	char buf[3], sts;
 	int i, err;
 
+	/*
+	 * initialize all members to -1 to signal the core that they are not
+	 * defined by the hardware.
+	 */
+	alm->time.tm_sec = -1;
+	alm->time.tm_min = -1;
+	alm->time.tm_hour = -1;
+	alm->time.tm_mday = -1;
+	alm->time.tm_mon = -1;
+	alm->time.tm_year = -1;
+	alm->time.tm_wday = -1;
+	alm->time.tm_yday = -1;
+	alm->time.tm_isdst = -1;
+
 	err = s35390a_get_reg(s35390a, S35390A_CMD_STATUS2, &sts, sizeof(sts));
 	if (err < 0)
 		return err;
